src/api/apiErrors.js

```diff
@@ -0,0 +1,62 @@
+/* @flow strict-local */
+import type { ApiResponseErrorData } from './transportTypes';
+import * as logging from '../utils/logging';
+
+/**
+ * Runtime class of custom API error types.
+ */
+export class ApiError extends Error {
+  data: $ReadOnly<ApiResponseErrorData>;
+  httpStatus: number;
+
+  constructor(httpStatus: number, data: $ReadOnly<ApiResponseErrorData>) {
+    super(data.msg);
+    this.data = data;
+    this.httpStatus = httpStatus;
+  }
+}
+
+/**
+ * Given a server response (allegedly) denoting an error, produce an Error to be
+ * thrown.
+ *
+ * If the `data` argument is actually of the form expected from the server, the
+ * returned error will be an {@link ApiError}; otherwise it will be a generic
+ * Error.
+ */
+export const makeErrorFromApi = (httpStatus: number, data: ?mixed): Error => {
+  // Validate `data`, and construct the resultant error object.
+  if (data !== null && typeof data === 'object') {
+    if (typeof data.msg === 'string' && data.result === 'error') {
+      if ('code' in data && typeof data.code === 'string') {
+        /* $FlowFixMe, in the unlikely event we can ever narrow types with if
+        statements on properties. */
+        return new ApiError(httpStatus, data);
```

Comment:
(Take the below as thoughts on a possible followup -- what's here is already an improvement, and will be a fine clean step in the history.)

Oh alternatively: one thing you could do here is say
```
        return new ApiError(httpStatus, { result: 'error', code: data.code, msg: data.msg });
```

That appears to pass Flow just fine.

I guess literally *that*, verbatim, isn't great because there are potentially those other properties on `data` that might be interesting. But it points in the direction of I think a pretty decent alternative: that version already cries out for ditching `result` as redundant, and pulling `code` and `msg` out onto the `ApiError` itself. Then the `data` property of `ApiError` could be just of type `{}`, any object at all; it could get either the original `data`, or perhaps better the result of a `const { result, code, msg, ...restData } = data`.

Reply:
Unfortunately, I'd better do this now.

There's a bug in my inlined implementation of `isEventQueueEtc.` that this will fix and Flow won't catch, because [caught exceptions are implicitly `any`, and _we can't turn that off_](https://github.com/facebook/flow/issues/2470).

Reply:
> [caught exceptions are implicitly `any`, and _we can't turn that off_](https://github.com/facebook/flow/issues/2470).

Womp womp.

Want to make a quick remark of that on `#mobile-team`? That will make it more likely we find this discussion if we see something that reminds us of it later... and also it's just an interesting fact for others to see.